File: LeetCode/Medium/0045-jump-game-ii/0045-jump-game-ii.java
class Solution {
    public int jump(int[] nums) {
        int n = nums.length;
        int[] dist = new int[n];
        for (int i = 0; i < n; i++) {
            dist[i] = Integer.MAX_VALUE;
        }
        
        Deque<Integer> q = new ArrayDeque<>();
        q.add(0);
        dist[0] = 0;

        while (q.size() > 0) {
            int cur = q.poll();

            for (int i = 1; i <= nums[cur]; i++) {
                int nxt = cur + i;
                if (nxt >= n) {
                    continue;
                }
                if (dist[nxt] > dist[cur] + 1) {
                    dist[nxt] = dist[cur] + 1;
                    q.add(nxt);
                }
            }
        }

        return dist[n-1];
    }
}

File: LeetCode/Medium/0045-jump-game-ii/0045-jump-game-ii.go
func jump(nums []int) int {
    n := len(nums)
    var stack []int
    dist := make([]int, n)

    for i := 0; i < n; i++ {
        dist[i] = math.MaxInt
    }

    stack = append(stack, 0)
    dist[0] = 0

    for len(stack) > 0 {
        top := len(stack) - 1
        cur := stack[top]
        stack = stack[:top]
        
        for j := 1; j <= nums[cur]; j++ {
            nxt := cur + j
            if (nxt >= n) {
                continue
            }
            if (dist[nxt] > dist[cur] + 1) {
                dist[nxt] = dist[cur] + 1
                stack = append(stack, nxt)
            }
        }
    }

    return dist[n-1]
}